### PR TITLE
Add distribution to warp using flat ID

### PR DESCRIPTION
**Pull request by ThomasRaoux**
This allows distribution to warps with a different shape than the block size definition.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp [hunk 1]
```diff
@@ -64,6 +65,179 @@ void transform_dialect::MapNestedForallToGpuThreadsOp::build(
...
+  }
+  IRMapping bvm;
+  for (auto [blockIdx, blockDim] :
+       llvm::zip(forallOp.getInductionVars(), threadMapping)) {
+    bvm.map(blockIdx,
+            threadOpsUpdated[blockDim.cast<DeviceMappingAttrInterface>()
+                                 .getMappingId()]);
+  }
+
+  // Step 4. Maybe create conditionals to predicate the region.
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:192]: I should finally be able to play with this today.

Quick offhand comment, can we refactor things upstream and avoid all the copy-pasta here and below ?

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:192]: It can go upstream if we are confident that it is what we need. I’m not fully sure this is the right logic until we plumb it through. What do you think?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp [hunk 2]
```diff
@@ -122,35 +296,34 @@ transform_dialect::MapNestedForallToGpuThreadsOp::applyToOne(
...
+      workgroupSize[0] * workgroupSize[1] * workgroupSize[2];
+  if (diag.succeeded() && (totalNumThreads % kWarpSize == 0)) {
+    target->walk([&](scf::ForallOp forallOp) {
+      // Ignore cases with different attributes.
+      for (Attribute map : forallOp.getMapping()->getValue()) {
+        if (!llvm::is_contained(warpMappingAttributes, map)) {
+          return WalkResult::skip();
+        }
+      }
+      if (diag.succeeded()) {
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:315]: I am having troubles with the nestings of diag.suceeded / walk / diag.succeeded.
Can we clean this a bit ?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/transform_distribute_forall.mlir [hunk 3]
```diff
@@ -13,38 +13,41 @@ hal.executable private @distribute {
...
-          {in_bounds = [true]} : vector<1xf16>, memref<1xf16, strided<[1], offset: ?>>
-        } {mapping = [#gpu.thread<x>]}
-// CHECK: %[[WX:.+]] = arith.divui %[[TX]], %[[C32]] : index
-// CHECK: vector.transfer_write %{{.*}}, %{{.*}}[%[[WX]]] {in_bounds = [true]} : vector<1xf16>, memref<1xf16, strided<[1], offset: ?>>
-        scf.forall (%arg0) in (%c8) {
-          vector.transfer_write %cst_0, %subview[%arg0]
-          {in_bounds = [true]} : vector<1xf16>, memref<1xf16, strided<[1], offset: ?>>
-        } {mapping = [#gpu.warp<x>]}        
+// CHECK:   vector.transfer_write %{{.*}}, %{{.*}}[%[[TY]], %[[TX]]] {in_bounds = [true]} : vector<1xf16>, memref<1024x1024xf16>
+        scf.forall (%arg0, %arg1) in (%c4, %c60) {
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/transform_distribute_forall.mlir:33]: You should now be able to just write: 
```
   scf.forall (%arg0, %arg1) in (4, 60) {
```
and drop the `arith.constant` genuflexions.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp [hunk 4]
```diff
@@ -64,6 +65,179 @@ void transform_dialect::MapNestedForallToGpuThreadsOp::build(
...
+  Value linearThreadId =
+      makeComposedAffineApply(rewriter, loc, tx + ty * BDX + tz * BDX * BDY,
+                              {threadIdX, threadIdY, threadIdZ});
+  return linearThreadId;
+}
+
+/// Rewrite scf.forall by distributing a flat id onto multiple dimensions.
+// This is a derived version of mapNestedForeachToThreadsImpl that handles flat
+// ids to support distributing with a different shape than the block sizes.
+static DiagnosedSilenceableFailure rewriteOneForallToGpuWithFlagThread(
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:100]: Is `FlagThread` meant to be `FlatThreadId` ?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp (hunk 1 above)
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:192]: ok but for review purposes, is `step 3.` the only variation compared to upstream ?

Can you mark what did not change with clear noisy blcks such as:
```
//=======================================//
// BEGIN: Copy-pasta from upstream, do not land
//=======================================//
...
//=======================================//
// END: Copy-pasta from upstream, do not land
//=======================================//
```

?

I'd like to know what I can ignore ..

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp [hunk 5]
```diff
@@ -122,35 +296,34 @@ transform_dialect::MapNestedForallToGpuThreadsOp::applyToOne(
...
+        }
+      }
+      if (diag.succeeded()) {
+        rewriter.setInsertionPoint(forallOp);
+        Value flatId = getFlatId(forallOp.getLoc(), rewriter, workgroupSize);
+        AffineExpr d0 = rewriter.getAffineDimExpr(0);
+        flatId = makeComposedAffineApply(rewriter, forallOp.getLoc(),
+                                         d0.floorDiv(kWarpSize), {flatId});
+        diag = rewriteOneForallToGpuWithFlagThread(
+            rewriter, forallOp, totalNumThreads / kWarpSize, flatId,
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:322]: Can we give proper name to `totalNumThreads / kWarpSize` and `flatId` ?
I do not know offhand what `flatSize` and `flatId` mean.

Are those actually `numWarps` and `threadIdWithinWarp` ? (given my understanding of the computation)

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp [hunk 6]
```diff
@@ -122,35 +296,34 @@ transform_dialect::MapNestedForallToGpuThreadsOp::applyToOne(
...
+    target->walk([&](scf::ForallOp forallOp) {
+      // Ignore cases with different attributes.
+      for (Attribute map : forallOp.getMapping()->getValue()) {
+        if (!llvm::is_contained(warpMappingAttributes, map)) {
+          return WalkResult::skip();
+        }
+      }
+      if (diag.succeeded()) {
+        rewriter.setInsertionPoint(forallOp);
+        Value flatId = getFlatId(forallOp.getLoc(), rewriter, workgroupSize);
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:317]: Can we call this `linearizedThreadId` ? This will carry the information that we obtained this id by linearizing.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp [hunk 7]
```diff
@@ -64,6 +65,179 @@ void transform_dialect::MapNestedForallToGpuThreadsOp::build(
   result.addTypes({pdl::OperationType::get(ctx)});
 }
 
+/// Return a flatten thread id for the workgroup with given sizes.
+static Value getFlatId(Location loc, RewriterBase &rewriter,
+                       ArrayRef<int64_t> workgroupSize) {
+  IndexType indexType = rewriter.getIndexType();
+  Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+  Value threadIdX =
+      workgroupSize[0] == 1
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:74]: are those `== 1` cases micro-optimizations that are not caught up by canonicalizations ?

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp [hunk 8]
```diff
@@ -64,6 +65,179 @@ void transform_dialect::MapNestedForallToGpuThreadsOp::build(
...
+  // newly created ops.
+  Value flatThreadId = flatId;
+  Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+  SmallVector<Value> threadOpsUpdated;
+  int64_t flatblockDim = 1;
+  for (int64_t dim : blockDims) {
+    flatblockDim *= dim;
+  }
+  int64_t dimUsed = 1;
+  for (int64_t dim : blockDims) {
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:169]: All this code looks like common linearization/delinearization in a basis for which we have facilities upstream (https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Utils/IndexingUtils.h#L96), why did you need to roll your own?

If you need more functionality please adapt.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp [hunk 9]
```diff
@@ -64,6 +65,179 @@ void transform_dialect::MapNestedForallToGpuThreadsOp::build(
   result.addTypes({pdl::OperationType::get(ctx)});
 }
 
+/// Return a flatten thread id for the workgroup with given sizes.
+static Value getFlatId(Location loc, RewriterBase &rewriter,
```
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp:69]: Note, this seems : https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Utils/IndexingUtils.h#L91
(we may want to update the upstream API a bit, builder as last argument is ugly, my bad ..)

**Review by nicolasvasilache — COMMENTED** [review]
So overall looking at the PR, I think I am seeing that it would be just a few minor lines of code to adapt upstream code using the proper upstream facilities. 

I am unclear yet if it is what we want as I have not yet connected it e2e but offhand, it seems reasonable (unless I got lost in all the code).

**Comment by ThomasRaoux** [comment]
This is not needed anymore based on latest improvements from @nicolasvasilache as far as I know

**Comment by nicolasvasilache** [comment]
indeed, this was generalized as part of https://reviews.llvm.org/D146130
